src/IRMatch.h

```diff
@@ -45,6 +51,2317 @@ bool expr_match(Expr pattern, Expr expr, std::map<std::string, Expr> &result);
+    s << "(" << op.a << " <= " << op.b << ")";
+    return s;
+}
+
+template<typename A, typename B>
+std::ostream &operator<<(std::ostream &s, const CmpOp<LT, A, B> &op) {
+    s << "(" << op.a << " < " << op.b << ")";
+    return s;
+}
+
+template<typename A, typename B>
+std::ostream &operator<<(std::ostream &s, const CmpOp<GE, A, B> &op) {
+    s << "(" << op.a << " >= " << op.b << ")";
+    return s;
+}
+
+template<typename A, typename B>
+std::ostream &operator<<(std::ostream &s, const CmpOp<GT, A, B> &op) {
+    s << "(" << op.a << " > " << op.b << ")";
+    return s;
+}
+
+template<typename A, typename B>
+std::ostream &operator<<(std::ostream &s, const CmpOp<EQ, A, B> &op) {
+    s << "(" << op.a << " == " << op.b << ")";
+    return s;
+}
+
+template<typename A, typename B>
+std::ostream &operator<<(std::ostream &s, const CmpOp<NE, A, B> &op) {
+    s << "(" << op.a << " != " << op.b << ")";
+    return s;
+}
+
+template<typename A, typename B>
+std::ostream &operator<<(std::ostream &s, const BinOp<Mod, A, B> &op) {
+    s << "(" << op.a << " % " << op.b << ")";
+    return s;
+}
+
+template<typename A, typename B>
+HALIDE_ALWAYS_INLINE
+auto operator+(A a, B b) noexcept -> BinOp<Add, decltype(pattern_arg(a)), decltype(pattern_arg(b))> {
+    return {pattern_arg(a), pattern_arg(b)};
+}
+
+template<typename A, typename B>
+HALIDE_ALWAYS_INLINE
+auto add(A a, B b) -> decltype(IRMatcher::operator+(a, b)) {return IRMatcher::operator+(a, b);}
+
+template<>
+HALIDE_ALWAYS_INLINE
+int64_t constant_fold_bin_op<Add>(halide_type_t &t, int64_t a, int64_t b) noexcept {
+    t.lanes |= ((t.bits >= 32) && add_would_overflow(t.bits, a, b)) ? MatcherState::signed_integer_overflow : 0;
+    int dead_bits = 64 - t.bits;
+    // Drop the high bits then sign-extend them back
+    return ((a + b) << dead_bits) >> dead_bits;
+}
+
+template<>
+HALIDE_ALWAYS_INLINE
+uint64_t constant_fold_bin_op<Add>(halide_type_t &t, uint64_t a, uint64_t b) noexcept {
+    uint64_t ones = (uint64_t)(-1);
+    return (a + b) & (ones >> (64 - t.bits));
+}
+
+template<>
+HALIDE_ALWAYS_INLINE
+double constant_fold_bin_op<Add>(halide_type_t &t, double a, double b) noexcept {
+    return a + b;
+}
+
+template<typename A, typename B>
+HALIDE_ALWAYS_INLINE
+auto operator-(A a, B b) noexcept -> BinOp<Sub, decltype(pattern_arg(a)), decltype(pattern_arg(b))> {
+    return {pattern_arg(a), pattern_arg(b)};
+}
+
+template<typename A, typename B>
+HALIDE_ALWAYS_INLINE
+auto sub(A a, B b) -> decltype(IRMatcher::operator-(a, b)) {return IRMatcher::operator-(a, b);}
+
+template<>
+HALIDE_ALWAYS_INLINE
+int64_t constant_fold_bin_op<Sub>(halide_type_t &t, int64_t a, int64_t b) noexcept {
+    t.lanes |= ((t.bits >= 32) && sub_would_overflow(t.bits, a, b)) ? MatcherState::signed_integer_overflow : 0;
+    int dead_bits = 64 - t.bits;
+    // Drop the high bits then sign-extend them back
+    return ((a - b) << dead_bits) >> dead_bits;
```

Comment:
This expr (and also the one on line 1694) is triggering UBSan: invalid-shift-base (presumably we are shifting negative or too-large amounts). Let me find out which.